Compiler/NFFrontEnd/NFInst.mo

```diff
@@ -1388,5 +1419,155 @@ algorithm
+          Error.addSourceMessageAndFail(Error.CLASS_EXTENDS_TARGET_NOT_FOUND, {name}, SCode.elementInfo(new));
+        end try;*/
+        ClassTree.Entry.CLASS(nodeExisting) := idExisting;
+        // get the existing class in the environment
+        existing := InstNode.definition(nodeExisting);
+        // TODO! FIXME! check if the base class (id) is replaceable and add an error if not
+        // Error.NON_REPLACEABLE_CLASS_EXTENDS
+        newName := name + "$ce$" + String(ceTick);
+        // patch the base class and the class extends!
+        (new, existing) := patchBaseClassAndClassExtends(newName, name, new, existing);
+        // update nodes!
+        newNode := InstNode.rename(newNode, newName);
+        newNode := InstNode.setDefinition(new, newNode);
+        nodeExisting := InstNode.setDefinition(existing, nodeExisting);
+
+        // do partial instantiation again TODO! FIXME! is this needed??!!
+        newNode := partialInstClass(newNode);
+        nodeExisting := partialInstClass(nodeExisting);
+
+        // replace the class extends with the new definition
+        scope := ClassTree.add(inScope, name, ClassTree.Entry.CLASS(nodeExisting), ClassTree.addConflictReplace);
+
+        // add the base class with the new name to the scope
+        scope := ClassTree.add(scope, newName, ClassTree.Entry.CLASS(newNode));
+      then
+        scope;
+
+    else inScope;
+  end match;
+end adaptClassExtendsChain;
+
+protected function patchBaseClassAndClassExtends
+  input String inBaseEltName;
+  input String inClassExtendsEltName;
+  input SCode.Element inBaseElt;
+  input SCode.Element inClassExtendsElt;
+  output SCode.Element outBaseElt;
+  output SCode.Element outClassExtendsElt;
+algorithm
+  (outBaseElt,outClassExtendsElt) := matchcontinue (inBaseElt,inClassExtendsElt)
+    local
+      SCode.Element elt,compelt,classExtendsElt;
+      SCode.Element cl;
+      SCode.ClassDef classDef,classExtendsCdef;
+      SCode.Partial partialPrefix1,partialPrefix2;
+      SCode.Encapsulated encapsulatedPrefix1,encapsulatedPrefix2;
+      SCode.Restriction restriction1,restriction2;
+      SCode.Prefixes prefixes1,prefixes2;
+      SCode.Visibility vis2;
+      String name1,name2,env_path;
+      Option<SCode.ExternalDecl> externalDecl1,externalDecl2;
+      list<SCode.Annotation> annotationLst1,annotationLst2;
+      SCode.Comment comment1,comment2;
+      Option<SCode.Annotation> ann1,ann2;
+      list<SCode.Element> els1,els2;
+      list<SCode.Equation> nEqn1,nEqn2,inEqn1,inEqn2;
+      list<SCode.AlgorithmSection> nAlg1,nAlg2,inAlg1,inAlg2;
+      list<SCode.ConstraintSection> inCons1, inCons2;
+      list<Absyn.NamedArg> clats;
+      list<tuple<SCode.Element, DAE.Mod, Boolean>> rest;
+      tuple<SCode.Element, DAE.Mod, Boolean> first;
+      SCode.Mod mods, derivedMod;
+      DAE.Mod mod1,emod;
+      SourceInfo info1, info2;
+      Boolean b;
+      SCode.Attributes attrs;
+      Absyn.TypeSpec derivedTySpec;
+
```

Comment:
From what I can see there's a lot of unneccesary code here. You don't seem to actually change the base class (and you shouldn't need to), so it's just taken apart and then reassembled unchanged. There also doesn't seem to be any need for two different cases, since it doesn't matter what kind the base class is.